SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,710 @@
+  var allDirectAndTransitiveUses: [Instruction] = []
+  
+  var conversionsAndReabstractions = ValueWorklist(context)                            
+  defer {
+    conversionsAndReabstractions.deinitialize()
+  }
+
+  var remainingUses = ValueWorklist(context)                            
+  defer {
+    remainingUses.deinitialize()
+  }
+
+  // A "root" closure undergoing conversions and/or reabstractions has additional restrictions placed upon it, inorder
+  // for a call site to be specialized against it. We handle conversion/reabstraction uses before we handle apply uses
+  // to gather the parameters required to evaluate these restrictions or to skip call site uses of "unsupported" 
+  // closures altogether.
+  //
+  // There are currently 2 restrictions that are evaluated prior to specializing a callsite against a converted and/or 
+  // reabstracted closure -
+  // 1. A reabstracted root closure can only be specialized against, if the reabstracted closure is ultimately passed
+  //    trivially (as a noescape+thick function) into the call site.
+  //
+  // 2. A root closure may be a partial_apply [stack], in which case we need to make sure that all mark_dependence 
+  //    bases for it will be available in the specialized callee in case the call site is specialized against this root
+  //    closure.
+
+  let haveUsedReabstraction = 
+    handleConversionsAndReabstractions(for: rootClosure, 
+                                       conversionsAndReabstractions: &conversionsAndReabstractions,
+                                       remainingUses: &remainingUses,
+                                       allDirectAndTransitiveUses: &allDirectAndTransitiveUses, context);
+
+
+  let intermediateClosureArgDescriptors: [IntermediateClosureArgDescriptor] = 
+    handleApplies(for: rootClosure, callSiteMap: &callSiteMap, remainingUses: &remainingUses, 
+                  allDirectAndTransitiveUses: &allDirectAndTransitiveUses, 
+                  convertedAndReabstractedClosures: &convertedAndReabstractedClosures,
+                  haveUsedReabstraction: haveUsedReabstraction, context)
+
+  finalizeCallSites(for: rootClosure, callSiteMap: &callSiteMap, allDirectAndTransitiveUses: allDirectAndTransitiveUses,
+                    intermediateClosureArgDescs: intermediateClosureArgDescriptors, context)
+}
+
+private func handleApplies(for rootClosure: SingleValueInstruction, callSiteMap: inout CallSiteMap, 
+                           remainingUses: inout ValueWorklist, allDirectAndTransitiveUses: inout [Instruction], 
+                           convertedAndReabstractedClosures: inout InstructionSet, haveUsedReabstraction: Bool, 
+                           _ context: FunctionPassContext) -> [IntermediateClosureArgDescriptor] 
+{
+  var intermediateClosureArgDescriptors: [IntermediateClosureArgDescriptor] = []
+  
+  while let use = remainingUses.pop() {
+    for use in use.uses {
+      allDirectAndTransitiveUses.append(use.instruction)
```

Comment:
Few questions/comments.
1. These missing forwarding instructions seem very similar to the conversions and reabstractions that we're handling?

2. Are `destroy_value`, `copy_value` and `move_value` it? Although I don't quite see why we'd need to handle `destroy_value` explicitly? 

3. I know you mentioned that we can support only OSSA in this pass. However, at least for Autodiff, the pass will need to be run twice -- once before the perf inliner (in the mid-level function pass pipeline) and once after it. Is it going to be possible to move the perf inliner pass before OSSA is eliminated as well?

Reply:
> These missing forwarding instructions seem very similar to the conversions and reabstractions that we're handling?

Yes, either you are handling those users in `handleApplies` or in `handleConversionsAndReabstractions`

> Are destroy_value, copy_value and move_value it?

I think so. If you see another kind of user, just add it.

> Although I don't quite see why we'd need to handle destroy_value explicitly?

All you need to do is to add it to the liferange.

Reply:
So I'm still a bit confused. Why is switching over different uses better than simply adding those uses to the `allTransitiveAndDirectUses` set without a switch? There seems to be no real difference here. Am I missing something?

Reply:
You can do that. The important thing is that you bail if you see an unknown user kind. Because that may let your closure life longer than you would compute by adding that user to the liferange.

Reply:
Alternatively you can compute the liferange separately with a utility like `NonEscapingClosureDefUseWalker`

Reply:
Sounds like that would be limited to non-escaping closures. Are you suggesting modifying it to support escaping closures as well? Or simply creating a def-use walker one for escaping closures?

Reply:
> But a user is a user, no? Unless the SIL is invalid, all uses of the closure must be valid; and the closure must itself be valid for all uses.

My point here is that the liferange of the closure doesn't necessarily end at a user. Let's take `move_value` as an example. You also need to add the user of the `move_value` to the liferange. If you just add all "immediate" users to the liferange, the resulting InstructionRange will not cover the real liferange of the closure

Reply:
Ok, so `move_value` is perhaps the only one we need to explicitly handle -- by appending the "new" value's uses to `allDirectAndTransitiveUses`. 

I get it. I'll modify things accordingly.

Reply:
Also keep in mind that we might add new instruction in future. The optimization should be bullet-prove and bail if it sees an unknown user kind

Reply:
The optimization should bail altogether or we bail that specific use and perhaps log about it.

Reply:
With "bail" I mean abort the optimization of this specific root closure.
Logging: Sure, you can `log` such an event (you will only see the log message if you set `verbose` to true)

Reply:
Ok